rust/sdk/src/client.rs

```diff
@@ -457,6 +458,684 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+    ///         println!("{:?}", event);
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// Generating an `OffsetMap` out of thin air is usually not possible because they
+    /// require stream IDs — which require knowledge of the streams and so on.
+    /// Hence, a more involved and useful example requires you to perform a query to
+    /// get an offset map when the query finishes streaming all results.
+    ///
+    /// ```no_run
+    /// use actyx_sdk::{service::QueryResponse, tags, Ax, AxOpts, Offset};
+    /// use futures::stream::StreamExt;
+    /// async fn upper_bound_example() {
+    ///     let service = Ax::new(AxOpts::default()).await.unwrap();
+    ///     // We're publishing events for a completely functional example
+    ///     let publish_response = service
+    ///         .publish()
+    ///         .event(
+    ///             tags!("temperature", "sensor:temp-sensor1"),
+    ///             &serde_json::json!({ "temperature": 10 }),
+    ///         ).unwrap()
+    ///         .event(
+    ///             tags!("temperature", "sensor:temp-sensor2"),
+    ///             &serde_json::json!({ "temperature": 21 }),
+    ///         ).unwrap()
+    ///         .event(
+    ///             tags!("temperature", "sensor:temp-sensor3"),
+    ///             &serde_json::json!({ "temperature": 40 }),
+    ///         ).unwrap()
+    ///         .await.unwrap();
+    ///     // Query for the "halfway" event
+    ///     let mut query_response = service
+    ///         .query("FROM 'sensor:temp-sensor2'")
+    ///         .await
+    ///         .unwrap();
+    ///     // This loop is a bit of a dirty hack for demonstration purposes
+    ///     // in real world usage you will most likely be using the events
+    ///     // and keeping the offset map in the end.
+    ///     let offsets = loop {
+    ///         let result = query_response.next().await.unwrap();
+    ///         if let QueryResponse::Offsets(offsets) = result {
+    ///             break offsets.offsets;
+    ///         }
+    ///     };
+    ///     // Query for all 'temperature' events with the previous query `OffsetMap`
+    ///     // as an upper bound. We're expecting to only see events after the "halfway"
+    ///     // event — {"temperature"}
+    ///     let mut query_response = service
+    ///         .query("FROM 'temperature'")
+    ///         .with_upper_bound(offsets.clone())
+    ///         .await.unwrap();
+    ///     while let Some(response) = query_response.next().await {
+    ///         println!("{:?}", response);
+    ///     }
+    /// }
+    /// ```
+    pub fn with_upper_bound(mut self, upper_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.upper_bound = Some(upper_bound);
+            return self;
+        }
+        panic!("Calling Query::with_upper_bound after polling.")
+    }
+
+    /// Dual to [`Query::with_upper_bound`], removes the upper bound.
```

Comment:
Please document what this means: the upper bound will be filled in by Actyx when processing the query, to be the currently known “present” offsets.